### PR TITLE
Feature to serve job directory files from archive location if available

### PR DESCRIPTION
**Pull request by tgianos**
The primary feature in this PR is to make it so if a job is complete (whether it was run via V3 API or V4 agent) and the job directory was archived the system can serve the directory resources from that archive location. This will enable us to continue serving job resources long after the logs have been deleted from disks or nodes have been replaced.

Other commits exist to support this change.

Jobs launched via V3 API call that are still in an active state will still have files transparently served off of disk on the node executing the job.

V4 agent jobs which are still active will currently return a 404 to all requests but eventually this will be supported by reverse proxy over the gRPC connection.

There are still several rough edges in the implementation in terms of handling errors and returning correct HTTP response codes however want to get this committed so it can be tested/iterated on.

### CODE REVIEW

File: genie-web/src/main/java/com/netflix/genie/web/controllers/ControllerUtils.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 package com.netflix.genie.web.controllers;
```
Comment (mprimi) [inline genie-web/src/main/java/com/netflix/genie/web/controllers/ControllerUtils.java]: Extra 'to' in the commit message?

File: genie-web/src/test/java/com/netflix/genie/web/controllers/ControllerUtilsUnitTests.java [hunk 2]
```diff
@@ -77,4 +77,25 @@ public void canGetRemainingPath() {
             .thenReturn("/api/v3/jobs/{id}/output/**");
         Assert.assertThat(ControllerUtils.getRemainingPath(request), Matchers.is("stdout"));
     }
+
```
Comment (mprimi) [inline genie-web/src/test/java/com/netflix/genie/web/controllers/ControllerUtilsUnitTests.java]: Maybe test corner cases just to freeze the behavior?
i.e. empty 'path', request not containing path, ...

File: genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/JobDirectoryManifest.java [hunk 3]
```diff
@@ -262,6 +263,10 @@ private ManifestEntry buildEntry(
             final boolean directory
         ) throws IOException {
             final String path = this.root.relativize(entry).toString();
+            final Path fileName = entry.getFileName();
+            final String name = fileName == null
```
Comment (mprimi) [inline genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/JobDirectoryManifest.java]: Should this error out instead of silently adding an entry with an empty name?

File: genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/JobDirectoryManifest.java [hunk 4]
```diff
@@ -319,6 +315,24 @@ private ManifestEntry buildEntry(
                 children
             );
         }
+
+        private String getMimeType(final String name, final Path path) {
+            // TODO: Loose conventions in here mostly for backwards compatibility but if there's empty
```
Comment (mprimi) [inline genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/JobDirectoryManifest.java]: We could/should move this `filename`->`mime type` map into configuration..

File: genie-web/src/main/java/com/netflix/genie/web/jpa/services/JpaJobPersistenceServiceImpl.java [hunk 5]
```diff
@@ -602,10 +583,48 @@ public void updateJobStatus(
...
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @Transactional(readOnly = true)
     public Optional<JobStatus> getJobStatus(@NotBlank(message = "Job id is missing and is required") final String id) {
+        // TODO: Need to decide on consistency of API. Do we throw exception? Return Optional? Runtime or Checked?
+        //       getJobArchiveLocation throws exception cause the underlying return type is already optional. So
+        //       do we return Optional<Optional<String>> that's really weird
         return this.jobRepository.findByUniqueId(id, JobStatusProjection.class).map(JobStatusProjection::getStatus);
```
Comment (mprimi) [inline genie-web/src/main/java/com/netflix/genie/web/jpa/services/JpaJobPersistenceServiceImpl.java]: The other two here are throwing `GenieNotFoundException`. What's the reason for not doing so for this one?

File: genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/JobDirectoryManifest.java (hunk 3 above)
Comment (tgianos) [reply genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/JobDirectoryManifest.java]: It was really put in to avoid the extreme edge case of the `Path` returning `null` and then calling `toString()` on it. The reason I went with empty string is cause that matches the behavior of the `path` attribute above where if you relativize the root to itself it returns empty string. The file name was really only used in this case for display in the UI so it's not critical.

File: genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/JobDirectoryManifest.java (hunk 4 above)
Comment (tgianos) [reply genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/JobDirectoryManifest.java]: That's a good idea! I'll add a TODO so can remember to come back to it

File: genie-web/src/main/java/com/netflix/genie/web/jpa/services/JpaJobPersistenceServiceImpl.java (hunk 5 above)
Comment (mprimi) [reply genie-web/src/main/java/com/netflix/genie/web/jpa/services/JpaJobPersistenceServiceImpl.java]: At least in this file, everything else is throwing `GenieJobNotFoundException`.
I know we need some consistency and a strategy at project level, but everything in this file is behaving the same,  I think having a single method behave differently for no apparent reason is more confusing, no?

File: genie-web/src/test/java/com/netflix/genie/web/controllers/ControllerUtilsUnitTests.java (hunk 2 above)
Comment (tgianos) [reply genie-web/src/test/java/com/netflix/genie/web/controllers/ControllerUtilsUnitTests.java]: Added another commit with some improved functionality
